lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -542,69 +536,103 @@ def deregister_image(module, connection):
                 connection.delete_snapshot(SnapshotId=snapshot_id)
         except botocore.exceptions.ClientError as e:
             # Don't error out if root volume snapshot was already deregistered as part of deregister_image
-            if e.error_code == 'InvalidSnapshot.NotFound':
+            if e.response['Error']['Code'] == 'InvalidSnapshot.NotFound':
                 pass
         exit_params['snapshots_deleted'] = snapshots
 
     module.exit_json(**exit_params)
 
 
-def update_image(module, connection, image_id, resource):
-    launch_permissions = module.params.get('launch_permissions') or []
-
-    if 'user_ids' in launch_permissions:
-        launch_permissions['user_ids'] = [str(user_id) for user_id in launch_permissions['user_ids']]
-
-    image = resource.Image(image_id)
-
+def update_image(module, connection, image_id):
+    launch_permissions = module.params.get('launch_permissions')
+    image = get_image_by_id(module, connection, image_id)
     if image is None:
         module.fail_json(msg="Image %s does not exist" % image_id, changed=False)
+    changed = False
+
+    if launch_permissions is not None:
+        current_permissions = image['LaunchPermissions']
+
+        current_users = set([permission['UserId'] for permission in current_permissions if 'UserId' in permission])
```

Comment:
These comprehensions don't need `[]` around them - `set(str(user_id) for user_id in launch_permissions.get('user_ids', []))` because 2.6+ handles comprehensions in function calls without them.